### PR TITLE
[AAP-38240] Migrate development environment to Podman

### PR DESCRIPTION
**Pull request by john-westcott-iv**
## Description

- Migrate local development, tox test containers, and image CI workflows to Podman.
- Retain Docker-named Make targets as compatibility aliases while using Podman underneath.
- Generate Podman-native host configuration, document local setup, and add version preflight checks.

## Type of Change

- [x] Development environment change
- [x] Configuration change
- [x] Documentation update
- [x] Test update

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have updated documentation where needed
- [x] I have considered error handling and edge cases
- [x] I have tested the changes in my local environment

### Prerequisites

- Podman 5.0 or later and podman-compose 1.6.0 or later.

### Steps to Test

1. Run make podman-preflight.
2. Run make tools/generated/sources and make podman-compose.
3. Run tox -e py312 and verify interrupted runs remove their managed Podman containers.

### Expected Results

- Generated service nodes use host.containers.internal.
- The development environment starts through Podman.
- Tox-managed containers are removed after normal and interrupted runs.

## Additional Context

- Legacy docker-compose Make targets remain aliases for compatibility.

## Summary by CodeRabbit

- **New Features**
  - Added Podman-based container workflows for development, testing, image building, and deployment.
  - Added automatic container health checks and cleanup during test runs.
  - Added support for multi-platform image builds targeting amd64 and arm64.

- **Documentation**
  - Updated setup, usage, testing, and troubleshooting guidance for Podman and `podman-compose`.
  - Clarified rootless versus rootful Podman proxy port behavior.

- **Bug Fixes**
  - Improved container readiness validation and service connectivity across local and CI environments.

### CODE REVIEW

File: .github/workflows/build-ci-image.yml [hunk 1]
```diff
@@ -32,14 +34,26 @@ jobs:
...
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.12'
+
+      - name: Install Podman
+        uses: redhat-actions/podman-install@072203a124c069d8fc8621c1a33f9fa9510ae2e1 # v1.1.1
+
+      - name: Install source-generation dependency
+        run: |
+          python -m pip install --only-binary :all: ansible-core==2.21.1
```
[the remark this answers is not in the document]
Comment (john-westcott-iv) [reply .github/workflows/build-ci-image.yml]: Fixed in b6f5687c: the workflow now installs pinned podman-compose==1.6.0 alongside ansible-core before make ci-image runs its Podman preflight.

File: tools/ansible/roles/sources/tasks/redis_cluster.yml [hunk 2]
```diff
@@ -27,7 +27,7 @@
 
     - name: Check cluster nodes
       shell:
-        cmd: "{{ docker_compose_cmd }} exec -t redis1 {{ redis_tls_enabled | ternary(base_redis_tls_command, base_redis_command) }} cluster nodes"
+        cmd: "{{ compose_cmd }} exec -T redis1 {{ redis_tls_enabled | ternary(base_redis_tls_command, base_redis_command) }} cluster nodes"
```
[the remark this answers is not in the document]
Comment (john-westcott-iv) [reply tools/ansible/roles/sources/tasks/redis_cluster.yml:30]: Fixed in b6f5687c: cluster nodes now retries up to 10 times with a 2-second delay until the command succeeds, while remaining unchanged.

File: toxfile.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+def published_port(name: str, exposed_port: ExposedPort) -> str:
+    result = run_podman("port", name, exposed_port.value)
+    address = result.stdout.strip().splitlines()[0]
+    return address.rsplit(":", 1)[1]
+
+
+def stop_container(name: str) -> None:
+    try:
+        run_podman("rm", "--force", name)
```
[the remark this answers is not in the document]
Comment (john-westcott-iv) [reply toxfile.py]: Fixed in b6f5687c: tox cleanup now gives podman rm --force a bounded 10-second timeout; the existing TimeoutExpired handling covers a blocked removal.

File: toxfile.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    ]
+    for key, value in config.environment.items():
+        args.extend(["--env", f"{key}={value}"])
+    for exposed_port in config.expose:
+        args.extend(["--publish", f"127.0.0.1::{exposed_port.value}"])
+    args.append(config.image)
+
+    LOGGER.warning("podman> run %s", config.image)
+    run_podman(*args)
+    RUNNING_CONTAINERS[tox_env.conf["env_name"]] = name
```
[the remark this answers is not in the document]
Comment (john-westcott-iv) [reply toxfile.py]: Fixed in b6f5687c: tox now retains a list of container names per environment and removes every tracked container during cleanup.

**Comment by jay-steurer** [comment]
I ran `make podman-compose` locally and iterated through the startup issues.

Findings:

- The initial preflight could not access the rootless Podman runtime from the sandbox; rerunning with host Podman access passed preflight with Podman 5.8.4 and podman-compose 1.6.0.
- Container startup then failed because the Makefile passed host UID `4208356`; rootless Podman could not map that UID/GID while initializing the `aap_gw_cprofiling` volume.
- Using UID `1000` allowed all five Jewel containers to start. I updated the local Makefile to select UID `1000` automatically for rootless Podman, while preserving an explicit `UID=...` override.
- The foreground attach also produced `crun exec.fifo` errors when `up --no-recreate` tried to restart already-running containers. The local Makefile now follows container logs directly instead.
- `make COMPOSE_UP_OPTS=-d podman-compose` completed successfully. The final health check returned HTTP 200 with `db_connected=true`, `proxy_connected=true`, and `dispatcherd_connected=true`.

No commit or push was made. The Makefile changes remain local and uncommitted.

File: Makefile [hunk 5]
```diff
@@ -6,25 +6,33 @@ CHECK_SYNTAX_FILES ?= aap_gateway_api/
...
-ifeq ($(CONTAINER_ENGINE),docker)
-CONTAINER_STARTUP_CONFIG := tools/configs/container-startup.yml
-else
-CONTAINER_STARTUP_CONFIG := tools/configs/container-startup-podman.yml
-endif
+PODMAN ?= podman
+PODMAN_COMPOSE ?= podman-compose --in-pod false
+PODMAN_MIN_VERSION ?= 4.9.3
+PODMAN_COMPOSE_MIN_VERSION ?= 1.6.0
+PODMAN_ROOTLESS := $(shell podman info --format '{{.Host.Security.Rootless}}' 2>/dev/null || true)
```
Comment (jay-steurer) [inline Makefile:13]: I reproduced this locally with rootless Podman. The current `UID := $(shell id -u)` passes host UID `4208356` into Compose, and startup then fails while initializing `aap_gw_cprofiling` with `potentially insufficient UIDs or GIDs available in user namespace` (`requested 4208356:0`). Please derive the Compose UID from `PODMAN_ROOTLESS` (defaulting to the image development UID `1000`, while retaining an explicit override) or use an equivalent rootless-safe mapping.

File: Makefile [hunk 6]
```diff
@@ -128,50 +128,88 @@ migrate-service-data:
...
+podman-compose-basic: $(SOURCES_STAMP) tools/generated/proxy.yml compose-build git_hooks_config
+	env UID=${UID} $(PODMAN_COMPOSE) -f tools/generated/compose.yml $(COMPOSE_OPTS) up --remove-orphans $(COMPOSE_UP_OPTS)
 
-## Start the docker container + plumb the sidecar containers and register services' proxy
-docker-compose: docker-compose-detached register-services plumb
+## Start the Podman containers, plumb sidecars, and register service proxies
+podman-compose: podman-compose-detached register-services plumb
 	@if [[ ! "${COMPOSE_UP_OPTS}" =~ "-d" ]] ; then \
-		env UID=${UID} $(DOCKER_COMPOSE) -f tools/generated/docker-compose.yml up --no-recreate; \
+		env UID=${UID} $(PODMAN_COMPOSE) -f tools/generated/compose.yml up --no-recreate; \
```
Comment (jay-steurer) [inline Makefile:138]: I also reproduced `up --no-recreate` emitting `crun exec.fifo` errors for already-running rootless containers after the detached startup had completed. The stack stayed healthy, but this masks an attach failure. Since the detached phase has already waited for readiness, use `podman-compose logs --follow` here to attach without attempting another start. With that change, `make COMPOSE_UP_OPTS=-d podman-compose` completed and the ping endpoint returned HTTP 200 with `db_connected=true`, `proxy_connected=true`, and `dispatcherd_connected=true`.

**Comment by tznamena** [comment]
## Test results

### Podman development stack

- ✅ `make podman-preflight` passed with Podman 5.8.4 and `podman-compose` 1.6.0.
- ❌ `make COMPOSE_UP_OPTS=-d podman-compose` fails under rootless Podman: the Makefile forwards the host UID, which may be outside the rootless user-namespace mapping during volume initialization.
- ✅ With `UID=1000` supplied as a workaround, the stack started, registration and plumbing completed, and both gateway and proxied ping returned HTTP 200 with `db_connected=true`, `proxy_connected=true`, and `dispatcherd_connected=true`.
- ❌ Foreground attachment after detached startup reproduces the `crun ... exec.fifo` errors with `podman-compose up --no-recreate`.
- ✅ Docker-named Make compatibility aliases resolve to Podman behavior.
- ✅ Test-created stack resources were cleaned up.

### tox and test-container lifecycle

- ✅ Full suite passed single-threaded: **2,339 passed, 10 xfailed, 3 xpassed** (~12m34s).
- ⚠️ The default parallel tox execution reached Podman startup but encountered four service-app startup errors. This needs investigation before relying on parallel tox as the standard workflow.
- ✅ On normal tox completion, the managed PostgreSQL container passed its health check and no `tox-podman-*` containers remained.
- ✅ On a forcibly interrupted tox run, the managed container reached readiness and the watchdog removed it; no `tox-podman-*` containers remained.

## Follow-up needed

The two open Makefile review threads are reproducible and should be addressed before approval:

1. Use a rootless-safe default Compose UID while preserving an explicit `UID=...` override.
2. Follow logs after detached startup rather than invoking `up --no-recreate` again.

Additionally, please investigate the four service-app startup failures from the default parallel tox run.

File: Makefile (hunk 5 above)
Comment (john-westcott-iv) [reply Makefile]: Fixed in 05a1d208. Rootless Podman now defaults the Compose UID to 1000, while an explicit `make UID=...` override remains supported.

File: Makefile (hunk 6 above)
Comment (john-westcott-iv) [reply Makefile]: Fixed in 05a1d208. The foreground phase now uses `podman-compose logs --follow` instead of a second `up --no-recreate`; its signal trap calls Compose down on Ctrl-C or termination to preserve the expected shutdown behavior.

**Review by jay-steurer — APPROVED** [review]
